File: interfaceui.py
from PySide6.QtCore  import QObject ,Slot,Signal,Property
from PySide6.QtQml import QQmlApplicationEngine, qmlRegisterType

class InterFaceUi(QObject):
    dataSignal = Signal(dict)
    updateSignal = Signal(int,dict)
    searchSignal = Signal(str)
    clearSearchSignal = Signal()
    removeAllSignal = Signal()
    removeSignal = Signal(int)
    def __init__(self,parent=None):
        super().__init__(parent)
        self._id = 0
        self._data = {}
    @Slot(str,str,str,str)
    def get_data(self,fname,lname,phone,email):
        self._id +=1
        dic = {"id":self._id,
               "FirstName":fname,
               "LastName":lname,
               "PhoneNumber":phone,
               "email":email}  
        self.dataSignal.emit(dic) 
    @Slot(int,str,str,str,str)
    def update(self,id,fname,lname,phone,email):
        
        dic = {"FirstName":fname,
               "LastName":lname,
               "PhoneNumber":phone,
               "email":email}  
        self.updateSignal.emit(id,dic)
    @Slot(str)
    def search(self ,name):
        self.searchSignal.emit(name)
    @Slot()
    def clearSearch(self ):
        self.clearSearchSignal.emit()
    @Slot()
    def removeAll(self ):
        self.removeAllSignal.emit()
    @Slot(int)
    def remove(self,id ):
        self.removeSignal.emit(id)

File: Cpp/interfaceui.cpp
#include "interfaceui.h"
#include <QDebug>
#include <mongomanager.h>
#include <listmodel.h>
InterfaceUI::InterfaceUI(MongoManager* mongo,ListModel * listmodel,QObject *parent)
    : QObject{parent},m_mongo{mongo},
    m_listmodel{listmodel}
{


}

void InterfaceUI::get_data(const QString &fname, const QString &lname, const QString &phone, const QString &email)
{
     bsoncxx::builder::stream::document doc{};
     doc << "id" << m_id
         << "FirstName" << fname.toStdString()
         << "LastName" << lname.toStdString()
         << "PhoneNumber" << phone.toStdString()
     << "email" << email.toStdString();

    m_mongo->addToDatabase(doc);
     QList<QVariantMap> listmap = m_mongo->data();
    QList<Item> listItem = convertMaptoItem(listmap);

    m_listmodel->refreshFromMongo(listItem);
    //qDebug() << fname << lname << phone << email;
}

void InterfaceUI::search(const QString &txt)
{
    QList <QVariantMap> list =  m_mongo->search(txt);
    QList<Item> listItem = convertMaptoItem(list);
    m_listmodel->refreshFromMongo(QList<Item>());
    m_listmodel->refreshFromMongo(listItem);
}

void InterfaceUI::clearSearch()
{
   QList<QVariantMap> list= m_mongo->data();
    QList<Item> Items =convertMaptoItem(list);
   m_listmodel->refreshFromMongo(Items);
}

void InterfaceUI::update( int id,const QString &fname, const QString &lname, const QString &phone, const QString &email)
{
    m_mongo->updateItem(id,fname,lname,phone,email);
    clearSearch();
}

void InterfaceUI::remove(int id)
{
    m_mongo->deleteItem(id);
    clearSearch();
}

void InterfaceUI::removeAll()
{
    m_mongo->deleteall();
    clearSearch();
}

QList<Item> InterfaceUI::convertMaptoItem(QList<QVariantMap> &list)
{

    QList<Item> listItem;

    foreach (QVariantMap var, list) {
        Item item;
        item.FirstName = var["FirstName"].toString();
        item.LastName = var["LastName"].toString();
        item.email = var["email"].toString();
        item.PhoneNumber = var["PhoneNumber"].toString();
        item.id = var["id"].toInt();
        listItem.append(item);
    }
    return listItem;
}
